File: week-5/word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        for i in range(len(board)):
            for j in range(len(board[i])):
                if board[i][j] == word[0] and self.dfs(i, j, 0, word, board):
                    return True
                
        return False
    
    def dfs(self, i, j, count, word, board):
        if count == len(word):
            return True
        if i<0 or j<0 or i>=len(board) or j>=len(board[i]) or board[i][j] != word[count]:
            return False
        
        temp = board[i][j]
        board[i][j] = ' '
        
        found = self.dfs(i+1, j, count+1, word, board) or self.dfs(i-1, j, count+1, word, board) or self.dfs(i, j+1, count+1, word, board) or self.dfs(i, j-1, count+1, word, board)
        
        board[i][j] = temp
        return found

File: week-5/word-search/word-search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function(board, word) {
    for (let i=0; i<board.length; i++) {
        for (let j=0; j<board[i].length; j++) {
            if (board[i][j] === word[0] && dfs(i, j, 0, word, board)) return true;
        }
    }
    
    return false;
};

const dfs = (i, j, ptr, word, board) => {
    if (ptr === word.length) return true;
    if (i<0 || j<0 || i>=board.length || j>=board[i].length || board[i][j] !== word[ptr]) return false;
    
    const temp = board[i][j];
    board[i][j] = ' ';
    const found = dfs(i+1, j, ptr+1, word, board) || dfs(i-1, j, ptr+1, word, board) || dfs(i, j+1, ptr+1, word, board) || dfs(i, j-1, ptr+1, word, board);
    
    board[i][j] = temp;
    return found;
};
